modules/navigation/navigation_geometry_parser_3d.h

```diff
@@ -0,0 +1,69 @@
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef NAVIGATION_GEOMETRY_PARSER_3D_H
+#define NAVIGATION_GEOMETRY_PARSER_3D_H
+
+#include "core/templates/vector.h"
+#include "scene/3d/node_3d.h"
+#include "scene/resources/navigation_mesh.h"
+
+class NavigationGeometryParser3D : public RefCounted {
+	GDCLASS(NavigationGeometryParser3D, RefCounted);
+	friend class NavigationMeshGenerator;
+
+protected:
+	static void _bind_methods();
+
+	GDVIRTUAL1RC(bool, _parses_node, Node *);
+
+	GDVIRTUAL4C(_parse_geometry, const Transform3D &, Node *, NavigationMesh::ParsedGeometryType, uint32_t);
+
+private:
+	Vector<float> *total_parsed_vertices = nullptr;
+	Vector<int> *total_parsed_indices = nullptr;
+
+	void parse_node_geometry(const Transform3D &p_navmesh_transform, Node *p_node, Vector<float> &p_vertices, Vector<int> &p_indices, NavigationMesh::ParsedGeometryType p_generate_from, uint32_t p_collision_mask);
+
+	void _add_vertex(const Vector3 &p_vec3);
+	void _add_mesh(const Ref<Mesh> &p_mesh, const Transform3D &p_xform);
+	void _add_mesh_array(const Array &p_array, const Transform3D &p_xform);
+	void _add_faces(const PackedVector3Array &p_faces, const Transform3D &p_xform);
+
+public:
+	virtual bool parses_node(Node *p_node);
+	virtual void parse_geometry(const Transform3D &p_navmesh_transform, Node *p_node, NavigationMesh::ParsedGeometryType p_generate_from, uint32_t p_collision_mask);
```

Comment:
`navmesh_transform` is the source transform from the start of the baking process. NavigationMesh resources have no transform so every offset needs to be baked. Maybe another term could make the parameter more clear?

Reply:
What I mean is, you used it in every single call to functions adding geometry, which suggests it's something the navmesh generator can do by itself, since not applying it would be wrong anyways. Is there any use case that would use it in any different way?
(I'm not referring to the transform of individual nodes providing geometries, these should still be provided)

Reply:
Properly the generator can handle that part, that was just how it was before. I only removed parameters from functions that didn't work with scripting, like passing the vertices and indices array's with every function call.

Reply:
In pseudocode:
```cpp
// Node X says "here's my geometry, and here is where I am"
add_arrays(arrays, transform)

// The geometry parser knows which navmesh it is processing,
// so it can combine its transform internally
void add_arrays(Arrays geometry, Transform3D geom_transform) {
	Transform3D total_transform = navmesh_transform * geom_transform;
	// Apply to geometry and add to navmesh
}
```